### DISCUSSION TITLE
Arduino\libraries\TFT_eSPI/Processors/TFT_eSPI_ESP32.h:14:25: fatal error: hal/gpio_ll.h: No such file or directory

### DISCUSSION THREAD

**Discussion by edwardcorke** [Bodmer/TFT_eSPI#3048] [General]
Can't compile due to no 'ha/gpio_ll.h' being found

I am trying to compile the colour_test example on an ESP32 Dev Module using the Arduino IDE. I've updated the user_setup.h to match the GC9A01 LCD I'm using. Please let me know if you have any suggestions for fixing this dependency issue.

Thanks!

**Comment by Bodmer** [comment]
Update to the latest ESP32 board package.

**Reply by Golowa70** [reply]
Hi. I have the same problem on VSCODE 1.85.1.    -   "pio\libdeps\ttgo-t1\TFT_eSPI\Processors/TFT_eSPI_ESP32.h:14:25: fatal error: hal/gpio_ll.h: No such file or directory".                                
Updates do not solve the problem. Please your help.

**Comment by jonsummers** [comment]
Found a work around that solved problem for my PlatformIO and ESP32 projects with TFT_eSPI V2.5.34.
Comment out the offending line in
[path]/TFT_eSPI/Processors/TFT_eSPI_ESP32.h
// #include "hal/gpio_ll.h"

This behaviour started with V2.5.34 (https://github.com/Bodmer/TFT_eSPI/releases/tag/V2.5.34) Nov 10, 2023, 
released Nov 17, 2023 when the offending line was added:
@@ -11,6 +11,7 @@
11 11    // Include processor specific header
12 12   #include "soc/spi_reg.h"
13 13   #include "driver/spi_master.h"
     _14 + #include "hal/gpio_ll.h"_
15 15

I don't know what problem it was trying to fix, but the **latest** ESP32 lib no longer has the folder hal within the ESP32 sdk.

**Reply by Bodmer** [reply]
You can read more about the history of this change here:
https://github.com/espressif/esp-idf/issues/9184

And the bug raised here for ESP EDF 5.1:
https://github.com/espressif/esp-idf/issues/11454

The latest Arduino board package 2.0.14  is based on ESP EDF 5.1 and the header does exist:
https://github.com/espressif/esp-idf/blob/master/components/hal/include/hal/gpio_hal.h

It's a pity the low level hal interface access keeps getting broken by the espressif software team, with no commitment to a stable low level hal.

**Comment by mintisan** [comment]
Maybe shoud update  espressif32 version as below:
```
[env:esp32dev]
platform = espressif32@6.5.0
```

**Comment by paulw517** [comment]
I have just encountered this same issue with TFT_eSPI/Processors/TFT_eSPI_ESP32.h working with a

Arduino IDE Ver: 1.8.12 and an Elecrow 3.5in ILI9488 TFT display with a ESP32-Wrover-B micro and Card Reader SKU: DIS05035H. 

When attempting to compile the sketch, the following compile returned the following:

C:\Users\Paul\Documents\Arduino\libraries\TFT_eSPI-master/Processors/TFT_eSPI_ESP32.h:14:10: fatal error: hal/gpio_ll.h: No such file or directory
 #include "hal/gpio_ll.h"

Screenshot(1199).png attached:

<img width="1920" height="1080" alt="Screenshot-library-User_Setup (1182)" src="https://github.com/user-attachments/assets/876071d4-799f-439a-8e88-ef8113d8c5d3" />

Any HELP with this issue would be greatly appreciated!

**Reply by Bodmer** [reply]
Make sure you have selected the ESP32 in the board menu. Make sure you have the latest board package loaded in the board manager:
<img width="779" height="213" alt="image" src="https://github.com/user-attachments/assets/1e3dcf6d-9e0d-466b-ba45-25b937ecdcdd" />

If interested you will find the file in the packages folder:
..... AppData\Local\Arduino15\packages\esp32\tools\esp32-libs\3.3.7\include\hal\esp32\include\hal
